Add websocket and mDNS support

jpcornil-git opened the pull request:
To support network access (wired or Wi-Fi) to Ayab shields and enable web application development, this PR proposes using the WebSocket protocol to transport serial data and adding mDNS (aka zeroconf/bonjour/...) to automatically discover "ayab" services/avoid the need to know/type IP addresses.

I believe the WebSocket protocol is a better approach than the low-level UDP/TCP proposal #727, as it provides a better option for web application development.

Description of changes:
- Addition of `mdns_discovery.py` (started by Engine and listening to ayab services)
- Addition of `websocketserial.py` (`WebsocketSerial` class with the same API as `Serial`)
- Serial port dropdown is now also populated with all `ayab` services discovered by mDNS (after a refresh)
- When selected, a `WebsocketSerial` object is created instead of a `Serial` 

This PR also add `utils/websocket2serial.py` to test this implementation with existing (serial) shield, e.g. it could run on a PC/RPi/... connected to the shield with USB Serial while ayab-desktop runs on a different machine/over internet/ .... This script also implements the mDNS service to enable automatic discovery.

I also prototyped an implementation on Uno R4 wifi (esp32 running an http/websocket server relaying serial data from/to the serial interface and mDNS announcing `ayab` service over the network)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * mDNS discovery of network devices; discovered boards appear in the port list and connect via WebSocket URLs.
  * New WebSocket-backed serial adapter and a WebSocket↔serial bridge for remote access.

* **Improvements**
  * Port list shows network names/IPs and updates safely from background discovery.
  * Request throttling to reduce repeated initialization attempts.

* **Bug Fixes**
  * Unified handling for local serial and network (WebSocket) connections.

* **Chores**
  * Added networking dependencies, updated CI packaging steps, and packaging args.

* **Style**
  * Minor formatting and readability tweaks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jpcornil-git committed:
Add a timeout mechanism for reqInfo retries

jpcornil-git committed:
Add support for websocket and mdns discovery

jpcornil-git committed:
Fix typing warnings/errors

jpcornil-git commented on requirements.txt:
```diff
@@ -6,3 +6,5 @@ types-pyserial==3.5.0
 types-mock
 types-Pillow==10
 mypy==1.9.0
+zeroconf==0.147.0
+websockets==15.0.1
```
I see 0.147.0 here https://pypi.org/project/zeroconf/

jpcornil-git commented on src/main/python/main/ayab/engine/communication.py:
```diff
@@ -93,7 +94,10 @@ def open_serial(self, portname: Optional[str] = None) -> bool:
             self.__portname = portname
             try:
-                self.__ser = serial.Serial(self.__portname, 115200, timeout=0.1)
+                if self.__portname and self.__portname[0:5] == "ws://":
+                    self.__ser = WebsocketSerial(self.__portname, timeout=0.1)
+                else:
+                    self.__ser = serial.Serial(self.__portname, 115200, timeout=0.1)
```
wss is not (yet) relevant here

jpcornil-git commented on src/main/python/main/ayab/engine/engine_fsm.py:
```diff
@@ -69,6 +71,15 @@ class StateMachine(QStateMachine):
     CONNECT: QState
     VERSION_CHECK: QState
 
+    lastRetry : float = 0.0
+
+    @staticmethod
+    def retry(method: Callable[..., None], args: Any=(), timeout: float=0.1) -> None:
+        current_time = time.time()
+        if (current_time - StateMachine.lastRetry) > timeout:
+            StateMachine.lastRetry = current_time
+            method(*args)
```
There is only a single instance of StateMachine (state methods are all static as well)

jpcornil-git commented on src/main/python/main/ayab/engine/websocketserial.py:
```diff
@@ -0,0 +1,42 @@
+    def write(self, data: bytes) -> None:
+        self._ws.send(data)
+
+    def close(self) -> None:
+        self._ws.close()
```
Agree/fixed

jpcornil-git commented on src/main/python/main/ayab/engine/websocketserial.py:
```diff
@@ -0,0 +1,42 @@
+    def in_waiting(self) -> int:
+        return len(self._rxbuffer)
+
+    def read(self, size: int = 1) -> bytes:
+        with self._lock:
+            while size > len(self._rxbuffer):
+                try:
+                    data = self._ws.recv(self._timeout)
+                    if isinstance(data, bytes):
+                        self._rxbuffer += data
+                except TimeoutError:
+                    break
+
+            data = self._rxbuffer[0:size]
+            self._rxbuffer = self._rxbuffer[size:]
+        return data
```
Only binary stream shouldbe considered here, text stream should be ignored (may be used for different purpose/out-og-band signaling moving forward).
Handling websockets.ConnectionClosed closed is a good suggestion/will fix

jpcornil-git commented on src/main/python/main/ayab/engine/mdns_discovery.py:
```diff
@@ -0,0 +1,113 @@
+            if name in self.services:
+                del self.services[name]
+
+    def _on_service_updated(self, zeroconf_instance: zeroconf.Zeroconf, type: str, name: str) -> None:
+        """Callback called when a service's information is updated."""
+        self.__logger.info(f"Service '{name}' of type '{type}' updated. Attempting to get new info...")
+        try:
+            # Re-fetch the service info as it might have changed
+            info = zeroconf_instance.get_service_info(type, name, timeout=3000)
+
+            if info:
+                self.__logger.info(f"Successfully retrieved updated info for service: {name}")
+                with self.services_lock:
+                    self.services[name] = info # Overwrite with new info
+            else:
+                self.__logger.warning(f"Failed to retrieve updated information for service {name} (it might have disappeared or changed unexpectedly).")
+        except Exception as e:
+            self.__logger.error(f"Error during service info update for {name}: {e}")
+
```
Correct, this is legacy code/will remove

jpcornil-git commented on src/main/python/main/ayab/engine/mdns_discovery.py:
```diff
@@ -0,0 +1,113 @@
+        def remove_service(self, zeroconf_instance: zeroconf.Zeroconf, type: str, name: str) -> None:
+            self.parent_browser._on_service_removed(zeroconf_instance, type, name)
+
+    def _on_service_changed(self, zeroconf_instance: zeroconf.Zeroconf, type: str, name: str, change_type: str) -> None:
+        """Callback called when a service is added or updated."""
+        info : zeroconf.ServiceInfo | None = zeroconf.ServiceInfo(type, name)
+        try:
+            info = zeroconf_instance.get_service_info(type, name, timeout=3000)
+            if info:
+                info = zeroconf_instance.get_service_info(type, name)
+                self.__logger.info(f"Service {change_type}: {name}")
+                with self.services_lock:
+                    self.services[name] = info
+            else:
+                self.__logger.warning(f"Unable to retrieve service {name} information")
+        except Exception as e:
+            self.__logger.error(f"Error while fetching service {name}: {e}")
+
```
Correct, I forgot to remove these line while refactoring that piece of code/will remove

jpcornil-git committed:
Address CodeRabbitAI review on cb56ebd

jpcornil-git committed:
Add dynamic update of the port combo for mDNS

jpcornil-git commented:
Just added support for dynamic update of the port combo for mDNS (and addressed CodeAabbitAI review)

jpcornil-git commented on src/main/python/main/ayab/engine/websocketserial.py:
```diff
@@ -0,0 +1,66 @@
+            self._rxbuffer = self._rxbuffer[size:]
+        return data
+
+    def write(self, data: bytes) -> None:
+        try:
+            with self._lock:
+                self._ws.send(data)
+        except websockets.ConnectionClosed:
+            raise ConnectionError("WebSocket connection is closed")        
+
```
Agree/will fix

jpcornil-git committed:
Change IP address used by websocket2serial.py for local IP detection

It was changed from 10.255.255 to 255.255.255.255 after SonarCould
security warning (not a real one as it is only used for detection of
the local IP address and not for communication ) but then socket
reports  localhost/127.0.0.1 as local ip address ...
=> Changed to 192.0.2.255 (not routable -> secure)

jpcornil-git committed:
Addressing coderabbitai formatting comments

jpcornil-git committed:
Add one extra argument to websocket2serial
Address a few more coderabbit formatting comments

jpcornil-git committed:
Fix handling of websocket path and multiple connections

jpcornil-git commented on utils/websocket2serial.py:
```diff
@@ -0,0 +1,250 @@
+        if ser and ser.is_open:
+            ser.close()
+            logging.info(f"Serial port {serial_port_name} closed for {websocket.remote_address}.")
+
```
Not an expected use case but I'll fix as well

jpcornil-git committed:
Fix requirements.*.txt

jonathanperret committed:
Add universal wheels for zeroconf and websockets

jonathanperret committed:
Add hidden imports to pyinstaller config

Workaround for https://github.com/pyinstaller/pyinstaller-hooks-contrib/issues/840

jpcornil-git committed:
Add zeroconf to pyinstaller hidden imports

jonathanperret committed:
Use collect-submodules rather than hidden_imports

Probably more resilient to changes in the zeroconf module

jpcornil-git committed:
Clean up, hook moved to base.json

jonathanperret commented on src/main/python/main/ayab/engine/engine.py:
```diff
@@ -73,6 +81,14 @@ def __init__(self, parent: GuiMain):
 
     def __del__(self) -> None:
         self.control.stop()
+        self.mdns_browser.stop()  # Lead to EventLoopBlocked exception when closing the window rather than the application
```
I'm reasonably confident you can remove that `__del__` method entirely — it fires too late anyway to do anything useful (Qt has already started tearing down) and as you noticed it can cause weird crashes.

jonathanperret replied:
In this case calling the parent class(es) destructors may make things even worse, as at the point where `__del__` is called (this happens only during app shutdown) Qt and the Python runtime have already started being torn down. Really, this destructor doesn’t serve any purpose (apart from sometimes triggering crashes on app exit) and should never have been added in the first place.

jpcornil-git replied:
It should either be completely removed (MdnsBrowser has it's own destructor) or the parent destructor should be called.
If you prefer the former, fine with me (not sure why control.stop() has to be called men Engine close)

jpcornil-git replied:
Do you want me to remove it now or do you handle that in a separate commit ?

jpcornil-git replied:
Done as well

jpcornil-git committed:
Move mDNS  before Simulation  in  the port dropdown menu

jpcornil-git committed:
Handle missing board_id from mDNS information

jonathanperret commented:
@jpcornil-git it looks like you didn't set up pre-commit hooks as indicated in the README: https://github.com/AllYarnsAreBeautiful/ayab-desktop/blob/main/README.md#running-from-source--development

Because of this, most of the files changed or added this PR have linter violations (lines too long, mostly). The CI workflow really should check for those but it doesn't currently (there's a `flake8` call in there but it uses different settings from `pre-commit` so it's not perfect yet), so what this means is that if we merged this PR as it it, the next poor soul who edited say `websocket2serial.py` would be faced with a large number of violations to fix.

You can run `pre-commit install` to set up the pre-commit hook, and run it on all the current files with `pre-commit run -a` to flush out any problems (it will auto-fix those it can). Note that this will flag a couple of violations that your PR doesn't introduce (e.g. a missing EOL in `bin/ayab`), feel free to either pick those auto-fixed changes into this PR or leave them aside.

jpcornil-git committed:
Address failing blake/flake8 precommit checks

jpcornil-git commented:
> @jpcornil-git it looks like you didn't set up pre-commit hooks as indicated in the README: https://github.com/AllYarnsAreBeautiful/ayab-desktop/blob/main/README.md#running-from-source--development

Correct/done now & changes checked in (I changed all files with failing checks, not only mine to help the next poor soul ;-)

jpcornil-git committed:
Remove Engine's destructor

jpcornil-git commented on src/main/python/main/ayab/engine/mdns_discovery.py:
```diff
@@ -0,0 +1,120 @@
+        with self.services_lock:
+            return self.services.copy()
+
+    def __del__(self) -> None:
+        """Call the stop method when the instance is destroyed."""
+        self.stop()
```
OK, I'll address in the next commit

jpcornil-git commented on src/main/python/main/ayab/engine/mdns_discovery.py:
```diff
@@ -0,0 +1,120 @@
+            self.__logger.error(f"Error while starting Zeroconf: {e}")
+            self.stop()
+
+    def stop(self) -> None:
+        """Stop Zeroconf service exploration and clean up."""
+        if not self.running:
+            self.__logger.warning("MdnsBrowser not running.")
+            return
+
+        self.__logger.info("Shutting down MdnsBrowser...")
+        if self.browser:
+            self.browser.cancel()
+        if self.zeroconf_instance:
+            self.zeroconf_instance.close()
+        self.running = False
+        self.__logger.info("ServiceBrowser stopped.")
```
OK, I'll address in the next commit

jpcornil-git commented on src/main/python/main/ayab/engine/mdns_discovery.py:
```diff
@@ -0,0 +1,120 @@
+        ) -> None:
+            self.parent_browser._on_service_removed(zeroconf_instance, type, name)
+
+    def _on_service_changed(
+        self,
+        zeroconf_instance: zeroconf.Zeroconf,
+        type: str,
+        name: str,
+        change_type: str,
+    ) -> None:
+        """Callback called when a service is added or updated."""
+        try:
+            info: zeroconf.ServiceInfo | None = zeroconf_instance.get_service_info(
+                type, name, timeout=3000
+            )
+            if info:
+                self.__logger.info(f"Service {change_type}: {name}")
+                with self.services_lock:
+                    self.services[name] = info
+                    self.update_callback()
+            else:
+                self.__logger.warning(f"Unable to retrieve service {name} information")
+        except Exception as e:
+            self.__logger.error(f"Error while fetching service {name}: {e}")
```
Good catch, likely a indentation fault, I'll rather move self.update_callback() out of the `with` block (should_notify unnecessary)

jpcornil-git commented on src/main/python/main/ayab/engine/mdns_discovery.py:
```diff
@@ -0,0 +1,120 @@
+        def __init__(self, parent_browser: "MdnsBrowser") -> None:
+            self.parent_browser = parent_browser
+
+        def add_service(
+            self, zeroconf_instance: zeroconf.Zeroconf, type: str, name: str
+        ) -> None:
+            self.parent_browser._on_service_changed(
+                zeroconf_instance, type, name, "added"
+            )
+
+        def update_service(
+            self, zeroconf_instance: zeroconf.Zeroconf, type: str, name: str
+        ) -> None:
+            self.parent_browser._on_service_changed(
+                zeroconf_instance, type, name, "updated"
+            )
+
+        def remove_service(
+            self, zeroconf_instance: zeroconf.Zeroconf, type: str, name: str
+        ) -> None:
+            self.parent_browser._on_service_removed(zeroconf_instance, type, name)
```
Agree, I'll address in the next commit

jpcornil-git committed:
Address coderabbit comments

jpcornil-git committed:
Use contextlib.suppress to improve readability

jpcornil-git commented on src/main/python/main/ayab/engine/engine.py:
```diff
@@ -114,12 +126,38 @@ def __activate_ui(self) -> None:
+        # Add serial ports
         utils.populate_ports(combo_box, port_list)
+        # Add websocket ports (collected by mDNS)
+        for _key, value in self.mdns_browser.get_known_services().items():
+            if value and value.server:
+                server_name = value.server.removesuffix(".local.")
+                try:
+                    server_ip = str(ipaddress.IPv4Address(value.addresses[0]))
+                except ipaddress.AddressValueError:
+                    server_ip = "unknown"
+                combo_box.addItem(f"{server_name} ({server_ip})", value)
         # Add Simulation item to indicate operation without machine
```
IPv6 addresses are not relevant here (`value` field is a `zeroconf.ServiceInfo` object and its `addresses` property is always IPv4, `parsed_addresses` should be used to include IPv6) but `addresses` property could indeed be missing if `__populate_ports()` is called juster after a new server record is registered/before IP address resolution, very unlikely but possible -> I'll fix in the next commit

jpcornil-git commented on src/main/python/main/ayab/engine/websocketserial.py:
```diff
@@ -0,0 +1,69 @@
+import websockets.sync.client
+import threading
+
```
Indeed, I'll fix in the next commit

jpcornil-git commented on utils/websocket2serial.py:
```diff
@@ -0,0 +1,351 @@
+    """
+    global serial_port_in_use
+
+    if websocket.request.path != WEBSOCKET_PATH:
+        error_message = (
+            f"Error: No handler for path {websocket.request.path}, "
+            f"closing connection"
+        )
+        await websocket.send(error_message)
+        logging.error(error_message)
+        return
```
This is not correct, websocket here is a ServerConnection object and connection shouldn't be closed here but will be closed when existing this handler by the server code.

jpcornil-git committed:
Various code improvements from coderabbitai review

jpcornil-git committed:
Remove /ws hardcoding in engine.py

Use the path from zeroconf properties instead.

jpcornil-git commented on src/main/python/main/ayab/engine/engine.py:
```diff
@@ -114,12 +126,35 @@ def __activate_ui(self) -> None:
+            self.__logger.info(f"Connecting to {board_id}")
+        else:
+            # Serial connection (default)
+            portname = self.ui.serial_port_dropdown.currentText()
```
Thanks for confirming, original code was therefore correct

jpcornil-git commented on src/main/python/main/ayab/engine/engine.py:
```diff
@@ -114,12 +126,36 @@ def __activate_ui(self) -> None:
 
     def __read_portname(self) -> str:
-        return self.ui.serial_port_dropdown.currentText()
+        # FIXME: method should return a class rather than a string to forward
+        # additional data
+        selected_index = self.ui.serial_port_dropdown.currentIndex()
+        user_data = self.ui.serial_port_dropdown.itemData(selected_index)
+        if user_data:
+            # Websocket connection
+            path = user_data.properties.get(b"path", b"/ws").decode()
+            board_id = user_data.properties.get(b"board_id", b"<Unknown>").decode()
+            portname = f"ws://{user_data.server.rstrip('.')}:{user_data.port}{path}"
+
+            self.__logger.info(f"Connecting to {board_id} at {path}")
+        else:
+            # Serial connection (default)
+            portname = self.ui.serial_port_dropdown.currentText()
+        return portname
```
This should be implemented once the server offers that (may never be required unless directly exposed over internet)

dl1com commented:
@jonathanperret @jpcornil-git merge fine for me - go?

jpcornil-git committed:
Code cleanup before merge

- Guard WebsocketSerial.in_waiting() with a lock
- Replace logger.error by logger.exception to get stack trace
- Replace rstrip by removesuffix to avoid removing valid characters in hostname
- Use startwith instead of equality to detect websocket URLs
- Remove unused arguments in functions
- Use contextlib.suppress to ignore exceptions when closing serial port

jpcornil-git commented:
I also tested on linux from my side (and tested on windows in the past -> I'll retest the latest build to make sure I didn't break anything with my cleanup :-)

jpcornil-git committed:
Address sonarcloud cosmetic comments

jpcornil-git commented on src/main/python/main/ayab/engine/engine.py:
```diff
@@ -114,12 +126,38 @@ def __activate_ui(self) -> None:
+            if value and value.server:
+                server_name = value.server.removesuffix(".local.")
+                try:
+                    server_ip = str(ipaddress.IPv4Address(value.addresses[0]))
+                except (IndexError, ipaddress.AddressValueError):
+                    server_ip = "unknown"
+                combo_box.addItem(f"{server_name} ({server_ip})", value)
```
This is probably not correct, see https://docs.python.org/3/library/ipaddress.html

jonathanperret commented:
I’m fine with merging as soon as JP is happy with the changes.

jpcornil-git commented on utils/websocket2serial.py:
```diff
@@ -0,0 +1,358 @@
+        logging.error(error_message)
+        return
+
+    if serial_port_lock.locked():
+        error_message = (
+            f"Error: Serial port {serial_port_name} is already in use by "
+            f"another connection."
+        )
+        await websocket.send(error_message)
+        logging.error(error_message)
+        return
+
```
Agree that we shold reject directly rather than queue/wait for a client timeout
=> I'll fix

jpcornil-git committed:
More code polish

- Make `WebSocketSerial.write` return the number of bytes written.
- MdnsBrowserstop(): Remove redundant check for running state.
- Remove redundant exception variable in logging.exception calls.
- Remove unecessary global declarations.

jpcornil-git committed:
Reject connection directly when already in use by another client

Previous implementation queued connection attempts, counting on client timeout.
This implementation try to acquire the lock with a timeout, and if it fails,
it sends an error message to the client and closes the connection immediately.

jpcornil-git commented on utils/websocket2serial.py:
```diff
@@ -0,0 +1,363 @@
+SERIAL_TIMEOUT = 0.1
+
+# --- WebSocket Configuration ---
+WEBSOCKET_HOST = "0.0.0.0"  # Listen on all available interfaces
+WEBSOCKET_PORT = 8080
+WEBSOCKET_PATH = "/ws"
```
That doesn't make sense for this application given that the client will always be elsewhere on the LAN a priori
